dlx2.py

```diff
@@ -0,0 +1,178 @@
+class Node:
+    def __init__(self, value, row, col):
+        self.value = value
+        self.row = row
+        self.col = col
+        self.up = self
+        self.down = self
+        self.right = self
+        self.left = self
+    
+    def add_above(self, node):
+        node.up = self.up
+        node.down = self
+        self.up.down = node
+        self.up = node
+
+    def add_below(self, node):
+        node.down = self.down
+        node.up = self
+        self.down.up = node
+        self.down = node
+
+    def add_right(self, node):
+        node.right = self.right
+        node.left = self
+        self.right.left = node
+        self.right = node
+
+    def add_left(self, node):
+        node.left = self.left
+        node.right = self
+        self.left.right = node
+        self.left = node
+
+    def __repr__(self) -> str:
+        return f'[{self.row}, {self.col}]'
+
+class Column:
+    def __init__(self, header):
+        self.head = header
+        self.tail = header
+        self.right = self
+        self.left = self
+        self.size = 1
+
+    def add(self, node):
+        node.up = self.tail
+        node.down = self.tail.down
+        self.tail.down = node
+        self.tail = node
+        self.head.up = node
+        self.size += 1
+
+    def __repr__(self):
+        string = ''
+        current_node = self.head
+        for i in range(self.size):
+            string += str(current_node)
+            current_node = current_node.down
+        return string
+
+class ColumnList:
+    def __init__(self, head_col):
+        self.head_col = head_col
+        self.tail_col = head_col
+        self.size = 1
+
+    def add(self, col):
+        col.right = self.tail_col.right
+        col.left = self.tail_col
+        self.tail_col.right = col
+        self.tail_col = col
+        self.head_col.left = col
+        self.size += 1
+
+    def __repr__(self):
+        string = ''
+        current_col = self.tail_col
+    
+        for i in range(self.size):
+            string += str(current_col)
+            string += '\n'
+            current_col = current_col.left
+        
+        return string
+
+# Class for a sudoku cover matrix 729 by 324
+# Each row represents a choice (e.g. 4 in cell 9, 5 in cell 80, 9 in cell 32, etc.)
+# Rows follow this order: cell 1 (guesses 1-9), cell 2 (guesses 1-9), etc.
+# Columns are used to eliminate conflicting choices
+# Columns 1-81 ensure that all other choices for a cell are eliminated after the first choice has been made
+# Columns 82-162 ensure that a digit may only be guessed once per sudoku row
+# Columns 163-243 ensure that a digit may only be guessed once per sudoku column
+# Columns 244-324 ensure that a digit may only be guessed once per box
+class Cover:
```

Comment:
good to see that you're documenting the structure of the cover matrix. Since this comment is describing the class, you could format this as a docstring:

```suggestion
class Cover:
    """Class for a sudoku cover matrix 729 by 324
    
    Each row represents a choice (e.g. 4 in cell 9, 5 in cell 80, 9 in cell 32, etc.)
    Rows follow this order: cell 1 (guesses 1-9), cell 2 (guesses 1-9), etc.
    Columns are used to eliminate conflicting choices
    Columns 1-81 ensure that all other choices for a cell are eliminated after the first choice has been made
    Columns 82-162 ensure that a digit may only be guessed once per sudoku row
    Columns 163-243 ensure that a digit may only be guessed once per sudoku column
    Columns 244-324 ensure that a digit may only be guessed once per box
    """
```

docstrings have some neat benefits:
* If you run `help(Cover)` in the console it'll print out this comment
* the string appears as `Cover.__doc__`
* documentation generators (e.g. Sphinx) will use this string.